keras_cv/visualization/plot_image_gallery.py

```diff
@@ -92,27 +89,88 @@ def unpackage_tfds_inputs(inputs):
             "to be true."
         )
 
-    fig = plt.figure(figsize=(cols * scale, rows * scale))
-    fig.tight_layout()  # Or equivalently,  "plt.tight_layout()"
-    plt.subplots_adjust(wspace=0, hspace=0)
-    plt.margins(x=0, y=0)
-    plt.axis("off")
+    def unpack_images(inputs):
+        return inputs["image"]
+
+    # Calculate appropriate number of rows and columns
+    if rows is None or cols is None:
+        if isinstance(images, tf.data.Dataset):
+            sample = next(iter(images.take(1)))
+            sample_shape = sample["image"].shape
+            if len(sample_shape) == 4:
+                batch_size = sample_shape[0]
+            else:
+                raise ValueError(
+                    "Passed `tf.data.Dataset` does not appear to be batched. Please batch using the `.batch().`"
+                )
+
+            images = images.map(unpack_images)
+            images = images.take(batch_size)
+            images = next(iter(images))
+        else:
+            sample_shape = images.shape
+            if len(sample_shape) == 4:
+                batch_size = sample_shape[0]
+            else:
+                raise ValueError(
+                    f"`plot_image_gallery` received unbatched images and `cols` and `rows` "
+                    "were both `None`. Either images should be batched, or `cols` and `rows` should be specified."
+                )
+
+    elif rows is not None and cols is not None:
+        if isinstance(images, tf.data.Dataset):
```

Comment:
If rows and cols are both specified, why do we need to compute batch size? I'm not sure what the purpose of this branch is

Reply:
This is done to take out the required number of images from the dataset

Reply:
I see. It seems like it would be a lot easier to just use `tfds.as_numpy(images.take(rows*cols)` 

(Probably have to optionally unbatch first)

Reply:
Overall, I think the logic would be much simpler if we separated the logic for converting a tf.data.Dataset into a numpy array from the logic for computing batch size / rows / cols. As-is there's some unnecessary indirection so the resulting code is more complex than it needs to be.

One simple approach would be to convert all possible input types to an unbatched iterator at the top of the method, and record the batch size if there was one.

Reply:
`tfds` is a special case. Most folks may also be invoking and creating `tf.data.Dataset` instances from a simple `image_dataset_from_directory()` call or `from_tensor_slices()` too. Using `tfds.as_numpy` might not be the best choice for that reason.